src/Analysis/Ast/Impl/Indexing/IndexManager.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using Microsoft.Python.Analysis.Core.Interpreter;
+using Microsoft.Python.Analysis.Documents;
+using Microsoft.Python.Core.IO;
+using Microsoft.Python.Parsing;
+
+namespace Microsoft.Python.Analysis.Indexing {
+    internal class IndexManager : IIndexManager {
+        private readonly ISymbolIndex _symbolIndex;
+        private readonly IFileSystem _fileSystem;
+        private readonly IndexParser _indexParser;
+        private readonly string _workspaceRootPath;
+        private readonly string[] _includeFiles;
+        private readonly string[] _excludeFiles;
+        private readonly ConcurrentDictionary<Uri, bool> _indexedFiles = new ConcurrentDictionary<Uri, bool>();
+
+        public IndexManager(ISymbolIndex symbolIndex, IFileSystem fileSystem, PythonLanguageVersion version, string rootPath, string[] includeFiles,
+            string[] excludeFiles) {
+            _symbolIndex = symbolIndex;
+            _fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
+            _indexParser = new IndexParser(symbolIndex, fileSystem, version);
+            _workspaceRootPath = rootPath ?? throw new ArgumentNullException(nameof(rootPath));
+            _includeFiles = includeFiles ?? throw new ArgumentNullException(nameof(rootPath));
+            _excludeFiles = excludeFiles ?? throw new ArgumentNullException(nameof(excludeFiles));
+        }
+
+
+        public void AddRootDirectory() {
+            var files = WorkspaceFiles();
+            foreach (var fileInfo in files) {
+                if (ModulePath.IsPythonSourceFile(fileInfo.FullName)) {
+                    Uri uri = new Uri(fileInfo.FullName);
+                    _indexParser.ParseAsync(uri);
```

Comment:
This code will spin up new task for every file and big amount of them will be executed in parallel. Since each task opens a synchronous file stream, that may create a heavy load on the file system. Would be good to test this method on something like ptvsd sources.
And please add `DoNotWait()` so that we don't have hanging tasks.

Reply:
This part is now an Async call too, but it will still spin up a new task for every file